### PR TITLE
Merging Jugalbandi Manager and PwR Studio setup

### PR DESCRIPTION
**Pull request by kanak8278**
* Combined docker compose file
* Combined postgres dump added
* Creating all topics in kafka
* Combined run.sh file added
* Port change for Jugalbandi Manager Frontend to 4179: Corresponding port changes in the Manager repository

Issues:
* ~~KAFKA_BROKER, KAFKA_USE_SASL variable in both environment Jugalbandi Manager and PwR Studio.~~
* PWR_SERVER_HOST vs SERVER_HOST in PwR Studio
* Two step run process: First run kafka and postgres and then again run all the services (Manager and Studio both) 
* Need to add docker-compose.stage.yml corresponding to the Jugalbandi Manager stage file.

### CODE REVIEW

**Comment by sameersegal** [comment]
Reference existing files across repositories rather than keeping a copy here. Why? Because those files will get updated with code changes. If you keep a copy here, it will get stale.


.env-dev = JBManager/.env-dev + PwR-Studio/.env-dev (minus Postgres & Kafka)
cat ../Jugalbandi-Manager/.env-dev > .env-dev
cat ../PwR-Studio/.env-dev | grep -R .... >> .env-dev

similarly from dump -- but remember there are two databases.

Similarly create-topic x2  (reference them)

**Comment by sameersegal** [comment]
Changed my mind. It will be very brittle if we try to achieve this by scripting. We should hardcode. If in the future, docker requirements change for any of the projects, we will manually make the change here.

Let's fix the image versions for JB Manager images.

**Comment by sameersegal** [comment]
Put all these files into a separate folder called `studio-and-manager` and add a README

File: .env-dev.template [hunk 1]
```diff
@@ -0,0 +1,77 @@
+SERVER_HOST=${JB_API_SERVER_HOST}
+
+KAFKA_BROKER=${JB_KAFKA_BROKER}
```
Comment (sameersegal) [inline .env-dev.template]: We have just KAFKA_BROKER and just POSTGRES_DATABASE_HOST

no prefixes JB_ and PWR_ required.

File: .env-dev.template [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+KAFKA_LANGUAGE_TOPIC=language
+KAFKA_FLOW_TOPIC=flow
+KAFKA_RETRIEVER_TOPIC=retriever
+KAFKA_INDEXER_TOPIC=indexer
+KAFKA_CHANNEL_TOPIC=channel
+
+POSTGRES_DATABASE_USERNAME=postgres
+POSTGRES_DATABASE_PASSWORD=postgres
+POSTGRES_DATABASE_HOST=${JB_POSTGRES_DATABASE_HOST}
+POSTGRES_DATABASE_NAME=postgres
```
Comment (sameersegal) [inline .env-dev.template]: Change DB name to jbmanager and jbstudio

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,295 @@
+services:
+  # Jugalbandi Manager
+  api:
+    build:
```
Comment (sameersegal) [inline docker-compose.yml]: Fix images to opennyai

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      - KAFKA_CFG_PROCESS_ROLES=controller,broker
+      - KAFKA_CFG_CONTROLLER_QUORUM_VOTERS=0@kafka:9093
+      # Listeners
+      - KAFKA_CFG_LISTENERS=PLAINTEXT://:9092,CONTROLLER://:9093,EXTERNAL://:19092
+      - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://kafka:9092,EXTERNAL://${JBHOST}:9092
+      - KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,EXTERNAL:PLAINTEXT
+      - KAFKA_CFG_CONTROLLER_LISTENER_NAMES=CONTROLLER
+      - KAFKA_CFG_INTER_BROKER_LISTENER_NAME=PLAINTEXT
+  # PwR Studio
+  server:
```
Comment (sameersegal) [inline docker-compose.yml]: Arrange these correctly:

JB Manager service
Studio service
db
kafka

(others are not required)

File: scripts/run.sh [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+
+# Create necessary directories
+# mkdir -p ./media
+mkdir -p ./server/dist
+
+# Determine host settings based on environment (WSL or Docker)
+if grep -qi microsoft /proc/version && grep -q WSL2 /proc/version; then
+    echo "Inside PwR Studio Configuration for WSL2"
+    JBHOST=$(hostname -I | awk '{print $1}')
```
Comment (sameersegal) [inline scripts/run.sh]: Clean up these. Use only JBHOST

Remove port number from JBHOST and set it correctly in .env file

File: scripts/run.sh [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    export POSTGRES_DATABASE_HOST="localhost"
+    echo "Setting hosts by docker-compose service names"
+fi
+
+# Set other variables
+export JB_API_SERVER_HOST="http://localhost:8000"
+export JB_ENGINE_VERSION
+
+# Build and run Docker Compose based on staging or development environment
+if [ -n "$stage" ]; then
```
Comment (sameersegal) [inline scripts/run.sh]: Remove stage flag

You need to build both frontends

**Review by sameersegal — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 3 above)
Comment (kanak8278) [reply docker-compose.yml]: opennyai images are giving error.
Same error across all the services
`flow-1      |     components["port"] = int(components["port"])
flow-1      |                          ^^^^^^^^^^^^^^^^^^^^^^^
flow-1      | ValueError: invalid literal for int() with base 10: 'None'`

File: studio-and-manager/.env-dev.template [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+PUBLIC_URL_PREFIX=
+
+# STORAGE_ACCOUNT_URL=
+# STORAGE_ACCOUNT_KEY=
+# STORAGE_AUDIOFILES_CONTAINER=audiofiles
+
+ENCRYPTION_KEY=
+
+WA_API_HOST=
+WA_API_KEY=
```
Comment (sameersegal) [inline studio-and-manager/.env-dev.template]: Is  this required?

WA_API_KEY and WA_API_ NUMBER

File: studio-and-manager/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,249 @@
+services:
+  # Jugalbandi Manager
+  api:
+    build:
+      context: ../../Jugalbandi-Manager
```
Comment (sameersegal) [inline studio-and-manager/docker-compose.yml]: Depend on uploaded images from DockerHub

File: studio-and-manager/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+  #   ports:
+  #     - "9000:9000"
+  kafka:
+    image: docker.io/bitnami/kafka:3.6
+    ports:
+      - "9092:19092"
+    volumes:
+      - "kafka_data:/bitnami"
+      - ./scripts/create-all-topics.sh:/usr/bin/create-topics.sh
+    command: "sh -c './run.sh & /usr/bin/create-topics.sh && tail -f /dev/null'"
```
Comment (sameersegal) [inline studio-and-manager/docker-compose.yml:224]: In the latest version of JB Manager, create-topic reads topic names from env values

File: studio-and-manager/scripts/create-all-topics.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+# Wait for Kafka to be ready
+while ! kafka_ready; do
+    echo "Waiting for Kafka to be ready..."
+    sleep 5
+done
+
+echo "Creating topics now"
+
+kafka-topics.sh --create --bootstrap-server localhost:9092 --topic channel --if-not-exists
```
Comment (sameersegal) [inline studio-and-manager/scripts/create-all-topics.sh]: Read topic names from env variables

File: studio-and-manager/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    ports:
+      - "5432:5432"
+    environment:
+      #   POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+    #   POSTGRES_DB: postgres
+    volumes:
+      - postgres_data:/var/lib/postgresql/data
+      - ./scripts/create-databases.sh:/docker-entrypoint-initdb.d/create-databases.sh
+      - ./jbmanager_dump.sql:/mnt/init/jbmanager_dump.sql
```
Comment (sameersegal) [inline studio-and-manager/docker-compose.yml]: Is the DB name specified in the dump.sql?

File: studio-and-manager/scripts/run.sh [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    export PWR_SERVER_HOST="http://api:3000"
+    export POSTGRES_DATABASE_HOST="localhost"
+    echo "Setting hosts by docker-compose service names"
+fi
+
+# Set other variables
+export JB_API_SERVER_HOST="http://localhost:8000"
+export JB_ENGINE_VERSION
+
+docker compose build "$@" --build-arg VITE_SERVER_HOST="$JB_API_SERVER_HOST"
```
Comment (sameersegal) [inline studio-and-manager/scripts/run.sh:71]: There are two frontends. Where is the second guys server host setting?

**Review by sameersegal — CHANGES_REQUESTED** [review]

**Review by sameersegal — APPROVED** [review]

**Pull request merged**